ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -301,6 +305,269 @@ prop_diffusion_target_established_local defaultBearerInfo diffScript =
+             "recent failures, opportunities, ignored too long)") $
+
+          signalProperty 20 show
+            (\(_,_,_,_,_,_,toolong) -> Set.null toolong)
+            ((,,,,,,) <$> (LocalRootPeers.toGroupSets <$> govLocalRootPeersSig)
+                     <*> govEstablishedPeersSig
+                     <*> govActivePeersSig
+                     <*> govActiveFailuresSig
+                     <*> trIsNodeAlive
+                     <*> promotionOpportunities
+                     <*> promotionOpportunitiesIgnoredTooLong)
+
+-- | A variant of
+-- 'Test.Ouroboros.Network.PeerSelection.prop_governor_target_active_local_above'
+-- but for running on Diffusion. This means it has to have in consideration the
+-- the logs for all nodes running will all appear in the trace and the test
+-- property should only be valid while a given node is up and running.
+--
+-- We do not need separate above and below variants of this property since it
+-- is not possible to exceed the target.
+--
+prop_diffusion_target_active_local_above :: AbsBearerInfo
+                                         -> DiffusionScript
+                                         -> Property
+prop_diffusion_target_active_local_above defaultBearerInfo diffScript =
+    let sim :: forall s . IOSim s Void
+        sim = diffusionSimulation (toBearerInfo defaultBearerInfo)
+                                  diffScript
+                                  tracersExtraWithTimeName
+                                  tracerDiffusionSimWithTimeName
+
+        events :: [Events DiffusionTestTrace]
+        events = fmap ( Signal.eventsFromList
+                      . fmap (\(WithName _ (WithTime t b)) -> (t, b))
+                      )
+               . Trace.toList
+               . splitWithNameTrace
+               . Trace.fromList ()
+               . fmap snd
+               . Signal.eventsToList
+               . Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . Trace.toList
+               . fmap (\(WithTime t (WithName name b)) -> (t, WithName name (WithTime t b)))
+               . withTimeNameTraceEvents
+                  @DiffusionTestTrace
+                  @NtNAddr
+               $ runSimTrace sim
+
+     in conjoin
+      $ verify_target_active_above
+      <$> events
+
+  where
+    verify_target_active_above :: Events DiffusionTestTrace -> Property
+    verify_target_active_above events =
+      let govLocalRootPeersSig :: Signal (LocalRootPeers.LocalRootPeers NtNAddr)
+          govLocalRootPeersSig =
+            selectDiffusionPeerSelectionState Governor.localRootPeers events
+
+          govActivePeersSig :: Signal (Set NtNAddr)
+          govActivePeersSig =
+            selectDiffusionPeerSelectionState Governor.activePeers events
+
+          -- Left () is a TrJoiningNetwork event
+          -- Right () is a TrKillingNode event
+          trJoinKillSig :: Signal (Either () ())
```

Comment:
Third place were we could use the same sum type; it seems like it's worth to do.

Reply:
Done